### PR TITLE
Improve and refactor Atom's light transmission in thin object mode

### PR DESCRIPTION
**Pull request by santipaprika**
The current implementation for translucency/light transmission through thin objects seems to be a bit buggy.

The proposed implementation (based on [Real-Time Realistic Skin Translucency](http://www.iryoku.com/translucency/downloads/Real-Time-Realistic-Skin-Translucency.pdf) by Jimenez et al) results in a realistic representation of translucency and offers multiple parameters to tune so that users can achieve the combination that best suits their scenario.

Here's a good-looking screenshot achieved with the proposed implementation:
image

The following video shows a quick comparison between the current implementation and a **work-in-progress version** of the proposed one (see the demo video below this one to check the final results).

https://user-images.githubusercontent.com/44426596/146233777-d4704c3c-06cd-4c13-b1bd-88d3a22cb014.mp4

  

Demo showing the **final results** and available parameters:

https://user-images.githubusercontent.com/44426596/146235417-1ea0606f-e87f-4f67-94ff-6dc3ca0099c0.mp4
  

As can be observed, this technique is highly dependent on the shadow maps resolution, which is the main cause of artifacts in some configurations. The solution for high cascade levels (for directional lights) or far disk/sphere lights has been to use a distance attenuation factor; if we do not attenuate the translucency contribution with distance, the artifacts due to low resolution will be very visible (see video).

In the following table, the influence of the shrink factor and angle offset can be seen more clearly:
thin-object-transmission-table-1
  
Signed-off-by: Santi Paprika <santi.gonzalez.cs@gmail.com>

### CODE REVIEW

**Review by HogJonny-AMZN — COMMENTED** [review]
Looks like a great improvement, these changes look good to me. Will leave review approval to engineers though.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lighting/LightingData.azsli
Comment (jeremyong) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lighting/LightingData.azsli:32]: In the transmission case, are the `shadowCoords` still used or are they replaced by these scaled shadow map queries? Can we alias the memory to keep this structure more compact if possible? (has a direct impact on tiled lighting occupancy)

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/DirectionalLight.azsli
Comment (jeremyong) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/DirectionalLight.azsli:91]: I suggest moving this and the attenuation parameter into `GetBackLighting` to reduce some code duplication and also make it easier to adjust the backlighting model in one place in the future

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/QuadLight.azsli
Comment (jeremyong) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/QuadLight.azsli:151]: Would it be better to simply add that constant within `GetBackLighting` by checking the transmission mode there instead of needing to perform this initialization for all light types?

**Review by jeremyong — COMMENTED** [review]
Thanks a lot for taking the time to implement this! I didn't write the original skin shader so I'll defer to another engineer there to supply additional comments, but I did leave some comments for you to consider.

Were you able to compare the before and after performance costs of the adjusted model?

File: Gems/Atom/Feature/Common/Assets/Materials/Types/EnhancedPBR.materialtype
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/Materials/Types/EnhancedPBR.materialtype:1253]: I think it would not be the same, since a sphere light shadow map (e.g.) will always have the same resolution, independently of the object's projected area into the shadow map (which is what we would like to take into account IMO, and what the distance attenuation factor approximately achieves).

Still, an additional attenuation factor could be derived from the shadow map resolution, I agree. It could avoid showing 'ugly' results at low-res configurations.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli:69]: Hm, yes, I grabbed this idea from the [reference paper](http://www.iryoku.com/translucency/downloads/Real-Time-Realistic-Skin-Translucency.pdf), but this will add energy (even though it can be compensated with the scale factor). Maybe a saturate would be more physically correct at the cost of clamping high-energy points to the same value? What do you think?
image

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/QuadLight.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/QuadLight.azsli]: You are right... I've changed this so that instead of initializing checking to the transmission mode, it initializes to `transmissionDistance = -1.0f`. Then, if shadows are not enabled, the variable won't be redefined and the `TransmissionMode::ThinObject` case in the `GetBackLighting` function will detect if `transmissionDistance < 0.0f`, and avoid computing the transmission in that case.
`ThickObject` transmission can be applied with shadows disabled, and in that case it does not matter if we use a 'null' value of 0.0f or -1.0f because the line `thickness = max(transmissionDistance, surface.transmission.thickness);` in the `GetBackLighting` `TransmissionMode::ThickObject` case will return the same (`surface.transmission.thickness` range is [0,1]).

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/DirectionalLight.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/DirectionalLight.azsli]: Makes sense! I've added a new parameter to the `GetBackLighting` function, `attenuationDistance` that contains the computed distance to use for attenuating (which might be different for each light type, e.g. distance to light in point or disk lights, and distance to the camera in directional lights).  
This way, the attenuation parameter computation can be moved into `GetBackLighting`. 
(done in https://github.com/o3de/o3de/pull/6428/commits/eda405f5f1d3ca785f61ea9ae1f4b9fb3ea54c0a)

File: Gems/Atom/Feature/Common/Assets/Materials/Types/EnhancedPBR.materialtype
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/Materials/Types/EnhancedPBR.materialtype]: Done in https://github.com/o3de/o3de/pull/6428/commits/46abff309336e54ff19507ef8f32c3a883f2df28.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lighting/LightingData.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lighting/LightingData.azsli]: Yes, it can be aliased (or we can also use a local variable, if needed in the future, see https://github.com/o3de/o3de/pull/6428/commits/6dbb8e1794f763302226dce41ff91e99b3b7e875). Aliasing done in https://github.com/o3de/o3de/pull/6428/commits/60ec8c3cfb5d9a1110a96977bdc3318739b3f1e0.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (Adi-Amazon) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli]: I think Jeremy has a point and since the albedo is multiplied outside the operation, you must saturate it.  The differences are very small in this example but you can still identify it and it does saturate more than it should.

image

On another note - great initiative and results

File: Gems/Atom/Feature/Common/Assets/Materials/Types/EnhancedPBR.materialtype
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/Materials/Types/EnhancedPBR.materialtype:1259]: nit: this might be a tad too close, especially for larger objects like sheet / hut cover (however it's a SoftMax, so the user can force it)

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli:27]: Could you add reference for that approximation??

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli:55]: nit: this might yield visual cutoff between the two curves.  Perhaps introduce smoothing?

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli:75]: as per remark from Jeremy - saturate the result

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli:78]: nit: use const for this magic number.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli:91]: nit: not sure that we need the sqrt and the power in this - both are expensive yet attenuation can be linear / cubic with good results.
Have you considered / tested?

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/CapsuleLight.azsli
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/CapsuleLight.azsli:139]: nit - remark about the parameters here (-1.0 that gets ignored if I follow your code properly)

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/LightTypesCommon.azsli
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/LightTypesCommon.azsli:52]: nit: explain the default parameters passed / possibly add them as const variable name for ease of understanding.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadow.azsli
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadow.azsli:159]: nice

**Review by Adi-Amazon — CHANGES_REQUESTED** [review]
Overall very good change that can add greatly to the visual.
Some minor alteration / consideration before submit.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli:27]: Done in https://github.com/o3de/o3de/pull/6428/commits/7052b99671027ceb36174d95e275d935eeb5be42. The reference is [GPU Gems 3 - 14. Advanced Techniques for Realistic Real-Time Skin Rendering](https://developer.nvidia.com/gpugems/gpugems3/part-iii-rendering/chapter-14-advanced-techniques-realistic-real-time-skin) (section 14.4.7)

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli]: Okay! Done in https://github.com/o3de/o3de/pull/6428/commits/c814be22fd65eacf47ea11d969acc8994780c2eb.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/CapsuleLight.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/CapsuleLight.azsli]: New explanations have been added as context to the default parameters (https://github.com/o3de/o3de/pull/6428/commits/69df812941fedec065e23e1965fc646f66a50624)

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/LightTypesCommon.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/LightTypesCommon.azsli]: Done in https://github.com/o3de/o3de/pull/6428/commits/69df812941fedec065e23e1965fc646f66a50624

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli]: Done in https://github.com/o3de/o3de/pull/6428/commits/69df812941fedec065e23e1965fc646f66a50624

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli:56]: This would only influence the thick object transmission mode, which might be out of the scope of this PR?  
The current O3DE implementation for thick objects hasn't been changed, only the name of the parameter `shadowRatio` for `transmissionDistance` in this case.  
Still, it could be done!

File: Gems/Atom/Feature/Common/Assets/Materials/Types/EnhancedPBR.materialtype
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/Materials/Types/EnhancedPBR.materialtype:1259]: In this case the higher the value the higher the attenuation with distance will be (it acts as a factor). Therefore, the factor can be 0 (or close to 0) for no attenuation no matter how far we are.  
Maybe I understood wrong... which is your suggestion?

**Comment by wintermute-motherbrain** [comment]
@antonmic Can you review this?

**Comment by hultonha** [comment]
Any update on this @Adi-Amazon @wintermute-motherbrain @antonmic, be awesome to get this in as it looks like a great improvement, just needs a couple more approvals. I can kick of a build when it's updated

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (Bindless-Chicken) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli:90]: Could you enable that by default through a preprocessor define for the skin material?

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/ProjectedShadow.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/ProjectedShadow.azsli]: Fixed in https://github.com/o3de/o3de/pull/6428/commits/1087e45ce3f8cb13669fd049957c6b52c9d653e9

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli]: Sure, makes sense! Done in https://github.com/o3de/o3de/pull/6428/commits/9c60bd008cbda990a237a54eaf71244cca141638.

**Comment by Bindless-Chicken** [comment]
My remarks have been addressed, this is good to go in my opinion.

**Review by Adi-Amazon — APPROVED** [review]
Very nice contribution!

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (antonmic) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli:78]: Comment says 
angle(N,L) + acos(transmissionNdLBias))
but calculation is
cos(angle(N,L) ) + transmissionNdLBias
Slightly confusing, would either leave the comment out or elaborate on the difference between the two (or am I missing something and the math works out to be the same?)

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/DirectionalLight.azsli
Comment (antonmic) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/DirectionalLight.azsli:81]: Why do we divide camToSurfDist by 2 here? Can you add a comment explaining?

**Review by antonmic — APPROVED** [review]
Nice work! A few small suggestions but otherwise ship it
Sorry for the delay in reviewing this

**Comment by rgba16f** [comment]
Branch is failing DCO check on some commits.

**Comment by rgba16f** [comment]
AR build started - https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-6428/1/

File: Gems/Atom/Feature/Common/Assets/Materials/Types/EnhancedPBR.materialtype
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/Materials/Types/EnhancedPBR.materialtype]: Right, thanks for remarking it! Done in https://github.com/o3de/o3de/pull/6428/commits/024933d4a177997467dc97271610dcb5f450363b.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/BackLighting.azsli]: The comment referred to the actual angle values. Since NdL = cos(angle(N,L)), then the angle added due to the bias would be acos(transmissionNdLBias), which is added to angle(N,L) to account for the total (extended) angle.  
I agree it was a bit confusing, and since the instruction in code is already self-explanatory, I removed it (https://github.com/o3de/o3de/pull/6428/commits/024933d4a177997467dc97271610dcb5f450363b).

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/DirectionalLight.azsli
Comment (santipaprika) [reply Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/PBR/Lights/DirectionalLight.azsli]: I thought it would make sense to attenuate the effect for different types of lights at different speeds, but after playing a bit with it, there's no big difference. Thanks for pointing out.

**Comment by jeremyong-az** [comment]
The last build aborted the gradle android build for some reason. Kicking it off again here: https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-6428/

**Comment by Bindless-Chicken** [comment]
Seems to have failed again in some physics test `test_ShapeCollider_CylinderShapeCollides`. Doesn't look related to the current change, but I'm not sure how to interpret the log.

**Comment by jeremyong-az** [comment]
Kicking it off again, I see an Asset Processor hangup, and I'm wondering if this is due to an instability that is resolved by this: https://github.com/o3de/o3de/pull/7289

**Pull request merged**
